# Car Price Prediction
## for an online car marketplace website in Serbia 2024

This paper is about understanding how prices of used cars are set up by analyzing numeric and categorical data of an online car marketplace website from Serbia. The paper follows the CRISP DM methodology

**Georgios Papadopoulos**  
Writer A  

**Peter Tolvaj**  
Writer B

# 1. Business Understanding

## Scenario and data source

The scenario involves a car listing marketplace website that connects buyers and sellers of used cars in Serbia. Although the website’s name is unknown, the website’s purpose is to efficiently match buyers who are looking for specific car attributes with sellers who want to sell their cars. It is irrelevant whether a seller or buyer is an individual person or a car dealership company that uses the website. Consequently, the key stakeholders include buyers, sellers and the website owner who acts as an intermediary and has the listing data available.

The marketplace faces the challenge of understanding what are the car attributes that lead to a sale. While listings receive views, it is unknown what attributes from the data source set up the price. By analyzing all listings, we will gather insight which the website owner can put into use for increasing the effectiveness of future car listings and determine if the price set up from the seller is acceptable / legitimate / not inflated, suggesting a price for the seller.

The data source is a csv file that provides detailed information on car listings collected from the marketplace website. The data source includes features that describe each listing with car attributes as globally known from the automotive industry (e.g. mileage, horsepower, fuel type). In addition, engagement metrics were scraped from the website in order to determine the views of its car listing and how many times a listed car was marked as favorite. More information about the data source will be given at the data provenance part.

## Business Objectives

We identified the following objectives that will directly address the implementation of our analysis.

- The first objective is to ensure that the marketplace website remains objective to connect buyer and seller with fair prices and operate accordingly. We want to avoid artificially inflate prices and thus, catch those prices that are not correlated.
- The second objective is to support the marketplace with data driven decision making. By analyzing patterns, such as which car attributes influence price, the marketplace website can focus on the listings with the features of the car that users actually care about to better meet their needs. 

## Business Success Criteria

The first part of the business success will be measured by the accuracy of the predictive price model. We assess this by looking at how well the model can predict future prices. Attributes like car type and horsepower play a key role, but also listings with more views and favorites indicate sellers are successfully attracting attention. Listings with complete details, such as car name and mileage, might perform better since they avoid missing values, and thus might have different pricing.

The second part relies on effective decision making. By analyzing which attributes, such as fuel type or engine capacity, correlate with price, the platform can refine its search and recommendation systems. This approach enables the platform to optimize the buyer experience by highlighting the most appealing attributes to users, helping sellers optimize their listings for better visibility and engagement, even without direct sales data.

# Data Mining Goals
The data mining goals for the marketplace website center on enhancing price predictability, understanding key buyer preferences and optimizing recommendations.

First, a primary goal is to predict future price levels for car listings based on key attributes such as car type, mileage, fuel type and year. This predictive model will allow the platform to optimize listing placement and visibility, ensuring that listings with high potential are showcased effectively. 

Another objective is to analyze buyer preferences by identifying which attributes consistently drive the most engagement, such as horsepower, color and engine capacity. Insights from this analysis will help guide sellers in crafting more appealing listings.

## Data Mining Success Criteria

The main success criteria found for the used car model is to determine the preference of the buyer. The buyer preference analysis should produce a reliable set of top 5 attributes that consistently attract buyers, apart from price.

## AI risks aspects

When building the price prediction model, we identified the following risks because prices can be volatile:

- **Model drift and accuracy issues:** The used car market changes constantly with shifting preferences, prices, and trends. If we don’t observe how the model behaves, it could drift from current conditions, leading to inaccurate recommendations or pricing.
- **Undetected biases:** The model could recommend certain car attributes, like popular brands, colors, or fuel types, based on historical data. This might lead to biased outputs, reducing visibility for less popular car types and limiting options for buyers. Sellers with niche offerings might face unfair treatment when e.g. selling expensive sport cars.


# 2. Data Understanding

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.float_format', '{:.0f}'.format) #to avoid scientific notation

serbia= pd.read_csv("serbia_car_sales_price_2024_v2.csv")
print(f"Dataset: {serbia.shape[0]} observations, {serbia.shape[1]} variables")

Dataset: 8413 observations, 18 variables


## Attribute types and semantics

Details of each attribute along with its type and a brief semantics description is given. The attribute type is already predefined based on the available metadata in JSON format which was then compared in R in case of inconsistencies. No inconsistencies were found.

During the analysis in the data understanding part, two new attributes were created by splitting up “horsepower” which are marked by asterisks (*) to be able to include horsepower in the correlation matrix from figure 1.

| Attribute | Type | Semantics |
|---|---|---|
| `views` | Integer | The total number of views the car listing has received. |
| `favorite` | Integer | The number of users who have added the car to their favorite list. |
| `post_info` | String | Contains information about the date when the announcement was created or updated since 08.10.2024. It may also include special notes provided by the seller. |
| `price` | Integer | The price of the car in Euro. |
| `car_name` | String | Manufacturer and car model. |
| `year` | Integer | Year of car manufacture. |
| `A/C` | String | Indicates whether the car is equipped with manual, automatic, or no air conditioning. |
| `emission_class` | String | The car’s emission standard classification, such as Euro 5. |
| `seats_amount` | Float | The number of seats in the car. |
| `horsepower` | String | The engine power measured in horsepower and kilowatts. |
| `color` | String | The exterior color of the car. |
| `car_mileage, km` | Float | The car’s mileage in kilometers. |
| `engine_capacity, cc` | Float | The engine capacity in cubic centimeters. |
| `type_of_drive` | String | Indicates the type of drive, such as front-wheel drive or all-wheel drive. |
| `doors` | String | The number of doors on the car. |
| `fuel` | String | The type of fuel the car uses, such as gasoline or diesel. |
| `car_type` | String | The category or body style of the car, such as sedan, SUV, or hatchback. |
| `gearbox` | String | The type of transmission, such as manual or automatic. |
| `hp*` | Float | The engine power measured in horsepower. |
| `kw*` | Float | The engine power measured in kilowatts. |

## Statistical properties

In order to define the statistical properties of the data source, the attributes were split up into numeric and categorical based on Table 1. 

### Numeric attributes

For the correlation analysis, a correlation matrix from figure 1 was decided so as to holistically calculate coefficients.  Figure 1 presents the correlation matrix of the numeric attributes of the data source. Strong positive correlations are observed between favorite and views with coefficient r=0.76 and between hp and kw = 1. These strong correlations are plausible, as a car listing gets more views, it is more likely to be marked as favorite. Likewise, hp (horsepower) and kw (Kilowatt) are both measurement units for describing engine power in different ways.

More interesting are the correlations between price and the other attributes, since the primary focus in the business understanding phase is price set as the target attribute. Positive correlations are observed between price and year with r=0.56, between price and engine capacity with r=0.24 and price and either horsepower or kilowatt with r=0.49. However, there is a weak negative correlation seen between price and car mileage with r=-0.02, meaning that the mileage of the used cars in Serbia does not directly affect pricing. 

In Table 3 all correlations were inputted in order to determine the significance of its attribute. The results show that the attributes favorite, year, engine capacity and horsepower/kilowatt are statistically significant in the given data source.

### Categorical attributes

For categorical attributes, the main focus would be to investigate the number of distinct values, missing values and the mode within each attribute. In Table 3 are computed the distinct and null values found in the data source. This step is needed to identify gaps that must be reported in the data quality section coming up.

The proportion of each categorical attribute will be analyzed and plotted at a later stage in the Visual Exploration phase.

## Data Quality

Regarding missing values, 13 out of 18 attributes contain missing values. Although most attributes contain missing values, they contain no more than 10 missing values based on table 2 and table 4. The only exception to that rule is the data source is the attribute emission_class with a total of 1337 missing values. Since the data source consists of 8413 entries, the missing values comprise 0.3% of the total data source, except emission class with 15.9% respectively. The potential effects are that the model to be built will be accurate, since the impact of missing values is negligible, imputation will not be necessary, thus no risk for bias will occur. In the data preparation report, we will dive deeper into the handling of missing values. When it comes to the attribute emission class, the reasons are not documented. We do not suspect any data entry errors. Nevertheless, as the first introduction of the Euro emissions classification started in July 1992 [1], the early car models in the data source might not have been officially classified, thus the missing values.

The information available on data provenance was given based on the JSON metadata file and the description of the data source, which both are limited. More specifically on provenance, data was collected from a popular web resource, which is not mentioned, where people sell and buy cars. The transformations were minimal and are translations from Serbian to English, removing announcements not related to selling cars and operations related to extraction of price and year [2]. 

In terms of plausibility of values, there were no inconsistencies found. The only concern that was raised during the analysis are the outliers in some attributes such as year, car mileage and engine capacity, which affect the distribution. The distribution of most attributes is rightly skewed due to the outliers. 


## Visual Exploration

The focus during the visual exploration was to identify whether there is any skewness, kurtosis about the data source and whether each attribute follows the normal distribution. The visual exploration was conducted without manipulating the data source, so we could identify what steps we would take in the data preparation chapter. In general, most visualizations are right skewed.

Figure 2 and Figure 3 represent the explorations we did, where we acknowledged that the numerical attributes are skewed and need to be examined further in part 3.


## Ethically Sensitive Attributes

The data source does not contain any ethically sensitive attributes. As the data source does not contain any personal data, such as but not limited to racial origin, personal finances and vulnerable entities [3]. The attributes views, favorite and post_info all are interaction data, for which they cannot be traced back to the user that created this interaction with any car listing.

## Risks and Biases

The only potential risk identified is related to the price attribute. More precisely, there might be a risk of market exploitation. Basically, sellers are artificially inflating the price of used car with the hope of profit maximization. Price volatility cannot be explained entirely through this model, since the given attributes are not enough to study the used cars market.

In terms of biases, confirmation bias is the first identified. Simple confirmations bias such toward low-mileage cars may reinforce the belief that they are always in better condition, even though the low-mileage cars might have been involved in car accidents, but also regular maintenance might make higher-mileage cars more reliable than low-mileage ones. Another bias could be when a user sees higher-priced listings for a specific car model, they may assume the car is worth more, even if the model suggests the prices are not linear to the model [4].

The questions that need to be addressed to any expert panel to analyze whether bias is present would be around confirmation bias and car experts. More specifically it needs to be addressed if the data source represents all used cars listed in Serbia and not only a subcategory of cars (e.g. luxurious cars), as some cars might not be within the price range of all socioeconomic brackets. Another question that needs to be addressed if whether the data source with the attribute car_mileage_km can be trusted, as Serbia has 10.3% used cars marked with mileage fraud, a tactic used to overvalue used cars by shrinking its mileage [5].


## Actions Required before Data Preparation

The actions required before data preparation in the third chapter have been identified as soon as the first statistical properties were introduced. The introduction of new attributes are needed, such as distinct engine power, separate horsepower and kilowatt, separated car brand and car model. In terms of data transformation, skewed attributes must be handled, as well as outlier and missing value removal. The final step would be to number the categorical attributes.
